### ISSUE TITLE
Restomm as an IMS GW - Support to proxy out SIP messages to IMS

### ISSUE THREAD

**Issue by gvagenas** [RestComm/Restcomm-Connect#1339]
Provide new feature that will allow UserAgentManager to proxy out SIP REGISTER messages and the next hop will be responsible to authenticate the user agent.
In that case Restcomm will act as a B2BUA and will route messages between the external registrar and the user agent.

If the registration is successful the Restcomm DB should be updated by using the register() method, so later Restcomm will be able to dial a registered client.

An example of how the UserAgentManager will be:

```
    @Override
    public void onReceive(final Object message) throws Exception {
        if (message instanceof ReceiveTimeout) {
            clean();
            keepAlive();
        } else if (message instanceof SipServletRequest) {
            final SipServletRequest request = (SipServletRequest) message;
            final String method = request.getMethod();
            if ("REGISTER".equalsIgnoreCase(method)) {
                if (imsGW) {
                    proxyRegisterToIms(); //This method is a B2BUA between the WebRTC client and IMS
                    return;
                }
                if(authenticateUsers) { // https://github.com/Mobicents/RestComm/issues/29 Allow disabling of SIP authentication
                    final String authorization = request.getHeader("Proxy-Authorization");
                    if (authorization != null) {
                      if (permitted(authorization, method)) {
                          register(message);
                      } else {
                          SipServletResponse response = ((SipServletRequest) message).createResponse(javax.servlet.sip.SipServletResponse.SC_FORBIDDEN); //Issue #935, Send 403 FORBIDDEN instead of issuing 407 again and again
                          response.send();
                      }
                    } else {
                        authenticate(message);
                    }
                } else {
                    register(message);
                }
            }
        } else if (message instanceof SipServletResponse) {
            SipServletResponse response = (SipServletResponse) message;
            if (response.getStatus()>400 && response.getMethod().equalsIgnoreCase("OPTIONS")) {
                removeRegistration(response);
            } else {
                pong(message);
            }
        }
    }

    private void proxyRegisterToIms() {
        //Here proxy to IMS
        //When 200 OK then call register()
        if (imsResponseCode==200) {
            register(request);
        }
    }
```

**Comment by gvagenas** [comment]
This closed with PR #1724

**Comment by gvagenas** [comment]
This closed with PR #1724
